python-lib/wordcloud_visualizer.py

```diff
@@ -0,0 +1,232 @@
+        plt.rcParams["axes.titlesize"] = self.titlesize
+        plt.title(title)
+        plt.imshow(wc, interpolation="bilinear")
+        return fig
+
+    @time_logging(log_message="Preparing data")
+    def _prepare_data(self, df: pd.DataFrame) -> list:
+        if self.subchart_column or self.language_column:
+            # Group data per language and subchart for tokenization
+            group_columns = [col for col in [self.language_column, self.subchart_column] if col]
+            df.dropna(subset=group_columns, inplace=True)
+            df_grouped = df.groupby(group_columns)
+        else:
+            # Simply format data similarly
+            df_grouped = [(self.language, df)]
+
+        return df_grouped
+
+    def _tokenize_texts(self, df_grouped: list) -> list:
+        """Tokenize each group of observations in its correct language"""
+        # Get language and subchart name for each group
+        texts = []
+        group_names = []
+        for name, group in df_grouped:
+            texts.append([group[self.text_column].str.cat(sep=" ")])
+            group_names.append(name)
+
+        # Get tokenization languages differently depending on language/subchart settings combinations
+        if not self.language_column and not self.subchart_column:
+            languages = [self.language]
+        elif self.language_column and self.subchart_column:
+            languages, subcharts = zip(*group_names)
+            languages = list(languages)
+            self.subcharts = list(subcharts)
+        elif self.subchart_column:
+            self.subcharts = group_names
+            languages = [self.language] * len(self.subcharts)
+        else:
+            languages = group_names
+
+        # Tokenize
+        docs = [self.tokenizer.tokenize_list(text, language)[0] for text, language in zip(texts, languages)]
+
+        return docs
+
+    @time_logging(log_message="Counting tokens")
+    def _count_tokens(self, docs: list):
+        """Count tokens in each group"""
+        counters = []
+        for doc in docs:
+            counter = Counter()
+            for token in doc:
+                counter[(token.text)] += 1  # Equivalently, token.lemma_
+            counters.append(counter)
+
+        if not self.subchart_column:
+            # sum the values with same keys
+            counts = Counter()
+            for d in counters:
+                counts.update(d)
+
+            counts = dict(counts)
+            return counts
+        else:
+            counts_df = pd.DataFrame(
+                list(zip(self.subcharts, counters)),
+                columns=["subchart", "count"],
+            )
+            counts_df = counts_df.groupby(by=["subchart"]).agg({"count": "sum"})
+            # remove subcharts emptied by filter
+            counts_df = counts_df.loc[counts_df["count"] != {}, :]
+            return counts_df
+
+    @time_logging(log_message="Generating wordclouds")
```

Comment:
Actually, now that this function yields instead of return, this logging decorator will always output 0 time. We should remove this and add logging in the recipe.py.
```suggestion
```